core/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java

```diff
@@ -160,27 +172,37 @@ else if (binders.size() > 1) {
 	}
 
 	private <T> Binder<T, ConsumerProperties, ProducerProperties> doGetBinder(String name, Class<? extends T> bindingTargetType) {
-		if (CollectionUtils.isEmpty(this.binderChildContextInitializers)) {
-			return this.doGetBinderConventional(name, bindingTargetType);
+		// If child initializers - use AOT lookup
+		if (!CollectionUtils.isEmpty(this.binderChildContextInitializers)) {
+			return this.doGetBinderAOT(name, bindingTargetType);
 		}
-		else {
-			if ((!StringUtils.hasText(name) || this.defaultBinder != null) && this.binderChildContextInitializers.size() == 1) {
+		return this.doGetBinderConventional(name, bindingTargetType);
+	}
+
+	private <T> Binder<T, ConsumerProperties, ProducerProperties> doGetBinderAOT(String name, Class<? extends T> bindingTargetType) {
+		// If neither name nor default given - return single or fail when > 1
+		if (!StringUtils.hasText(name) && !StringUtils.hasText(this.defaultBinder)) {
+			if (this.binderChildContextInitializers.size() == 1) {
 				String configurationName = this.binderChildContextInitializers.keySet().iterator().next();
+				this.logger.debug("No specific name or default given - using single available child initializer '" + configurationName + "'");
 				return this.getBinderInstance(configurationName);
 			}
-			else if (this.defaultBinder != null && this.binderChildContextInitializers.size() > 1) {
```

Comment:
This improves the existing impl and now behaves as follows (inline comments hopefully describe this as well):
- if no name or no default specified then fail if > 1 binders or return single binder if 1 binder available
- if name specified be sure it exists in child initializers (else fail)
- if default specified be sure it exists in child initializers (else fail)
- else fail